glean/lang/lsif/Data/LSIF/Types.hs

```diff
@@ -0,0 +1,315 @@
+  -- These are glass-like symbol ids, identified by opaque text and a scheme
+  | LsifMoniker {
+      monikerKind :: !MonikerKind,
+      scheme :: !Text,
+      identifier :: !Text
+  }
+  | PackageInformation {
+      name :: !Text,
+      manager :: !Text,
+      version :: !Text
+  }
+  | SymbolRange {
+      range :: {-# UNPACK #-}!Range,
+      tag :: !Tag
+  }
+  | ResultSet -- indirection nodes
+  | DefinitionResult
+  | DeclarationResult
+  | ReferenceResult
+
+  | LsifDiagnosticResult {
+      diagnostics :: Vector Diagnostic
+  }
+  | LsifDocumentSymbolResult {
+      documentSymbols :: Vector Id
+  }
+  | LsifUnknown
+  --
+  -- Edge constructors
+  -- Edges are labels + an in and out vertex id
+  --
+  | Edge {
+      label :: !EdgeLabel,
+      outV :: !Id,
+      inV :: !Id
+    }
+  | Contains {
+    outV :: !Id,
+    inVs :: Vector Id
+  }
+  | Item {
+    outV :: !Id,
+    inVs :: Vector Id,
+    document :: !Id,
+    property :: Maybe Property
+  }
+  deriving (Generic, Show, Eq)
+
+data HoverContents
+  = HoverSignature {
+      hoverLanguage :: Text,
+      value :: Text
+  }
+  | HoverText Text
+  deriving (Eq, Show)
+
+data Diagnostic
+  = Diagnostic {
+      severity :: !Int,
+      code :: !Int,
+      message :: Text,
+      diagnosticRange :: {-# UNPACK #-}!Range
+  }
+  deriving (Eq, Show)
+
+data MonikerKind = Export | Local
+  deriving (Show, Eq)
+
+data Marker = Begin | End
+  deriving (Show, Eq)
+
+data Scope = DocumentScope | ProjectScope
+  deriving (Show, Eq)
+
+-- | A 0-indexed line/character-offset point in a document
+data Position = Position {
+      line :: {-# UNPACK #-} !Int64,
+      character :: {-# UNPACK #-} !Int64
+    }
+  deriving (Generic, Eq, Show)
+
+data Range = Range {
+      start :: {-# UNPACK #-} !Position,
+      end :: {-# UNPACK #-} !Position
+    }
+  deriving (Generic, Eq, Show)
+
+-- | LSIF tags, very close to LSP method call results
+data Tag
+  = Definition {
+    -- The text covered by the range
+      tagText :: !Text,
+    -- The kind of the declaration.
+      tagKind :: !SymbolKind,
+      -- The full range of the declaration not including leading/trailing
+      -- whitespace but everything else, e.g comments and code.  The range must
+      -- be included in fullRange.
+      fullRange :: {-# UNPACK #-}!Range,
+      -- Optional detail information for the declaration.
+      tagDetail :: Maybe Text
+  }
+  | Declaration {
+    -- The text covered by the range
+      tagText :: !Text,
+    -- The kind of the declaration.
+      tagKind :: !SymbolKind,
+      -- The full range of the declaration not including leading/trailing
+      -- whitespace but everything else, e.g comments and code.  The range must
+      -- be included in fullRange.
+      fullRange :: {-# UNPACK #-}!Range,
+      -- Optional detail information for the declaration.
+      tagDetail :: Maybe Text
+  }
+  | Reference {
+      tagText :: !Text
```

Comment:
this is the key for looking up the Definition/Declaration? Should it be a `newtype` maybe?